src/module_wrap.cc

```diff
@@ -494,23 +493,109 @@ Maybe<uv_file> CheckFile(const URL& search,
   return Just(fd);
 }
 
+PackageConfig emptyPackage = { false, false, "", false };
+std::unordered_map<std::string, PackageConfig> pjson_cache_;
+
+PackageConfig& GetPackageConfig(Environment* env, const std::string path) {
+  auto existing = pjson_cache_.find(path);
+  if (existing != pjson_cache_.end()) {
+    return existing->second;
+  }
+  Maybe<uv_file> check = CheckFile(path, LEAVE_OPEN_AFTER_CHECK);
+  if (check.IsNothing()) {
+    return pjson_cache_[path] = emptyPackage;
+  }
+
+  Isolate* isolate = env->isolate();
+  Local<Context> context = isolate->GetCurrentContext();
+  std::string pkg_src = ReadFile(check.FromJust());
+  uv_fs_t fs_req;
+  uv_fs_close(nullptr, &fs_req, check.FromJust(), nullptr);
+  uv_fs_req_cleanup(&fs_req);
+
+  // It's not okay for the called of this method to not be able to tell
+  // whether an exception is pending or not.
+  TryCatch try_catch(isolate);
+
+  Local<String> src;
+  if (!String::NewFromUtf8(isolate,
+                           pkg_src.c_str(),
+                           v8::NewStringType::kNormal,
+                           pkg_src.length()).ToLocal(&src)) {
+    return pjson_cache_[path] = emptyPackage;
+  }
+
+  Local<Value> pkg_json;
+  if (!JSON::Parse(context, src).ToLocal(&pkg_json) || !pkg_json->IsObject())
+    return (pjson_cache_[path] = emptyPackage);
+  Local<Value> pkg_main;
+  bool has_main = false;
+  std::string main_std;
+  if (pkg_json.As<Object>()->Get(context, env->main_string())
+                              .ToLocal(&pkg_main) && pkg_main->IsString()) {
+    has_main = true;
+    Utf8Value main_utf8(isolate, pkg_main.As<String>());
+    main_std = std::string(*main_utf8, main_utf8.length());
+  }
+
+  bool esm = false;
+  Local<Value> pkg_mode;
+  std::string pkg_mode_std;
+  if (pkg_json.As<Object>()->Get(context, env->mode_string())
+                              .ToLocal(&pkg_mode) && pkg_mode->IsString()) {
+    Utf8Value pkg_mode_utf8(isolate, pkg_mode.As<String>());
+    pkg_mode_std = std::string(*pkg_mode_utf8, pkg_mode_utf8.length());
+    if (pkg_mode_std == "esm") {
+      esm = true;
+    }
+  }
+
+  PackageConfig pjson = { true, has_main, main_std, esm };
+  return pjson_cache_[path] = pjson;
+}
+
```

Comment:
If this C++ json reader/parser is faster than JS would it be possible in a follow up PR to just use it for reading of package.json's or at least cached the parsed data here to avoid a subsequent load/parse?

Reply:
Yes the goal is to unify on a single cache with further work.